sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -214,6 +214,8 @@ class CodeGenContext {
     case dt: DataType if isPrimitiveType(dt) => s"($c1 > $c2 ? 1 : $c1 < $c2 ? -1 : 0)"
     case BinaryType => s"org.apache.spark.sql.catalyst.util.TypeUtils.compareBinary($c1, $c2)"
     case NullType => "0"
+    case s: StructType if s.fields.forall(_.dataType.isInstanceOf[AtomicType]) =>
```

Comment:
I think you hit a wall here... One way to do this is to introduce someway to add a function directly to the generated class, and then call that function. Basically introduce a new function in CodeGenContext that takes in some code and a function name.